model/lifecycle.go

```diff
@@ -1080,25 +1083,131 @@ func setNumDeps(tasks []*task.Task) {
 	for i, task := range tasks {
 		idToTask[task.Id] = tasks[i]
 	}
-
-	for _, task := range tasks {
+	deduplicatedTasks := []*task.Task{}
+	for _, task := range idToTask {
+		task.NumDependents = 0
+		deduplicatedTasks = append(deduplicatedTasks, task)
+	}
+	for _, task := range deduplicatedTasks {
 		// Recursively find all tasks that task depends on and increments their NumDependents field
 		setNumDepsRec(task, idToTask, make(map[string]bool))
 	}
 }
 
 // setNumDepsRec recursively finds all tasks that task depends on and increments their NumDependents field.
 // tasks not in idToTasks are not affected.
-func setNumDepsRec(task *task.Task, idToTasks map[string]*task.Task, seen map[string]bool) {
-	for _, dep := range task.DependsOn {
+func setNumDepsRec(t *task.Task, idToTasks map[string]*task.Task, seen map[string]bool) {
+	for _, dep := range t.DependsOn {
 		// Check whether this dependency is included in the tasks we're currently creating
-		if depTask, ok := idToTasks[dep.TaskId]; ok {
-			if !seen[depTask.Id] {
-				seen[depTask.Id] = true
-				depTask.NumDependents = depTask.NumDependents + 1
-				setNumDepsRec(depTask, idToTasks, seen)
-			}
+		depTask, ok := idToTasks[dep.TaskId]
+		if !ok {
+			// it's possible to get in to this block if tasks depend on others outside of the task's
+			// version, such as with commit queue merges. This scenario is currently not handled here
+			// because commit queue merges are all linked lists, and this block would do nothing
```

Comment:
Do you mean that it's okay we aren't recursing because commit queue merges will only have dependency depth of 1?

Reply:
I don't see how this logic is any different from how the logic already was, except that you added a comment. Before we only added to seen if `ok && !seen[depTask.Id]` and that's also the current behavior.

Reply:
yea sorry this block is actually what caused the most problems. Adequately explaining it requires me drawing a picture which I can do if you're interested, but the short version is that I previously had some extra code in here to robustly handle scenarios that (I believe) can't actually happen with the current system, which panicked. I removed it but left the structure here with a comment

Reply:
the drawing a picture part was for the first question :) I'll try to make the comment more succinct